crates/subspace-service/src/mmr/request_handler.rs

```diff
@@ -0,0 +1,297 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::mmr::get_offchain_key;
+use futures::channel::oneshot;
+use futures::stream::StreamExt;
+use parity_scale_codec::{Decode, Encode};
+use sc_client_api::{BlockBackend, ProofProvider};
+use sc_network::config::ProtocolId;
+use sc_network::request_responses::{IncomingRequest, OutgoingResponse};
+use sc_network::{NetworkBackend, PeerId};
+use schnellru::{ByLength, LruMap};
+use sp_core::offchain::storage::OffchainDb;
+use sp_core::offchain::{DbExternalities, OffchainStorage, StorageKind};
+use sp_runtime::codec;
+use sp_runtime::traits::Block as BlockT;
+use std::collections::BTreeMap;
+use std::hash::{Hash, Hasher};
+use std::marker::PhantomData;
+use std::sync::Arc;
+use std::time::Duration;
+use tracing::{debug, error, trace};
+
+const MAX_NUMBER_OF_SAME_REQUESTS_PER_PEER: usize = 2;
+
+/// Defines max items per request
+pub const MAX_MMR_ITEMS: u32 = 20000;
+
+mod rep {
+    use sc_network::ReputationChange as Rep;
+
+    /// Reputation change when a peer sent us the same request multiple times.
+    pub const SAME_REQUEST: Rep = Rep::new(i32::MIN, "Same state request multiple times");
+}
+
+/// Generates a `RequestResponseProtocolConfig` for the state request protocol, refusing incoming
+/// requests.
+pub fn generate_protocol_config<
+    Hash: AsRef<[u8]>,
+    B: BlockT,
+    N: NetworkBackend<B, <B as BlockT>::Hash>,
+>(
+    _: &ProtocolId,
+    genesis_hash: Hash,
+    fork_id: Option<&str>,
+    inbound_queue: async_channel::Sender<IncomingRequest>,
+) -> N::RequestResponseProtocolConfig {
+    N::request_response_config(
+        generate_protocol_name(genesis_hash, fork_id).into(),
+        Vec::new(),
+        1024 * 1024,
+        16 * 1024 * 1024,
+        Duration::from_secs(40),
+        Some(inbound_queue),
+    )
+}
+
+/// Generate the state protocol name from the genesis hash and fork id.
+pub fn generate_protocol_name<Hash: AsRef<[u8]>>(
+    genesis_hash: Hash,
+    fork_id: Option<&str>,
+) -> String {
+    let genesis_hash = genesis_hash.as_ref();
+    if let Some(fork_id) = fork_id {
+        format!(
+            "/{}/{}/mmr/1",
+            array_bytes::bytes2hex("", genesis_hash),
+            fork_id
+        )
+    } else {
+        format!("/{}/mmr/1", array_bytes::bytes2hex("", genesis_hash))
+    }
+}
+
+/// The key of [`BlockRequestHandler::seen_requests`].
+#[derive(Eq, PartialEq, Clone)]
+struct SeenRequestsKey {
+    peer: PeerId,
+    starting_position: u32,
+}
+
+#[allow(clippy::derived_hash_with_manual_eq)]
+impl Hash for SeenRequestsKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.peer.hash(state);
+        self.starting_position.hash(state);
+    }
+}
+
+/// Request MMR data from a peer.
+#[allow(clippy::derive_partial_eq_without_eq)]
```

Comment:
Why suppressing these instead of deriving `Eq` as well? Similarly for other clippy suggestions that should be addressed instead of suppressed.